src/DataValues/TimeValue.php

```diff
@@ -136,29 +136,35 @@ public function __construct( $timestamp, $timezone, $before, $after, $precision,
 	 * @throws IllegalValueException
 	 * @return string
 	 */
-	private function validateIsoTimestamp( $timestamp ) {
+	private function normalizeIsoTimestamp( $timestamp ) {
 		if ( !is_string( $timestamp ) || $timestamp === '' ) {
 			throw new IllegalValueException( '$timestamp must be a non-empty string' );
 		} elseif ( !preg_match(
-			'/^([-+])(\d{1,16})-(\d\d)-(\d\d)(T(?:[01]\d|2[0-3]):[0-5]\d:(?:[0-5]\d|6[012])Z)$/',
+			'/^([-+])(\d{1,16})-(\d\d)-(\d\d)T(\d\d):(\d\d):(\d\d)Z$/',
 			$timestamp,
 			$matches
 		) ) {
 			throw new IllegalValueException( '$timestamp must resemble ISO 8601, given ' . $timestamp );
 		}
 
-		list( , $sign, $year, $month, $day, $time ) = $matches;
+		list( , $sign, $year, $month, $day, $hour, $minute, $second ) = $matches;
 
 		if ( $month > 12 ) {
 			throw new IllegalValueException( 'Month out of allowed bounds' );
 		} elseif ( $day > 31 ) {
 			throw new IllegalValueException( 'Day out of allowed bounds' );
+		} elseif ( $hour > 23 ) {
```

Comment:
I should have benchmarked this before. The bottleneck is the regex (~100ms for 50000 strings). The difference between the two regexes is not measurable. The additional comparison (no matter if the type cast is done implicitly via `$hour > 23` or explicitly via `(int)$hour`) adds ~10ms. `intval` should be avoided since it's ~5 times slower.